### PR TITLE
expression: migrate tests to testify

### LINKED ISSUE #29132: migrate test-infra to testify for `expression/distsql_builtin_test.go`

### LINKED ISSUE #29134: migrate test-infra to testify for `expression/expr_to_pb_test.go`

### LINKED ISSUE #29142: migrate test-infra to testify for `expression/schema_test.go`

### LINKED ISSUE #29164: data race in `expression/builtin_cast_vec_test.go`
## Bug Report

Please answer these questions before submitting your issue. Thanks!
```
[2021-10-27T02:27:43.503Z] WARNING: DATA RACE
[2021-10-27T02:27:43.503Z] Read at 0x00c000320000 by goroutine 94:
[2021-10-27T02:27:43.503Z]   math/rand.(*rngSource).Uint64()
[2021-10-27T02:27:43.503Z]       /usr/local/go/src/math/rand/rng.go:239 +0x3e
[2021-10-27T02:27:43.503Z]   math/rand.(*rngSource).Int63()
[2021-10-27T02:27:43.503Z]       /usr/local/go/src/math/rand/rng.go:234 +0x1d9
[2021-10-27T02:27:43.503Z]   math/rand.(*Rand).Int63()
[2021-10-27T02:27:43.503Z]       /usr/local/go/src/math/rand/rand.go:85 +0xbc
[2021-10-27T02:27:43.503Z]   math/rand.(*Rand).Float64()
[2021-10-27T02:27:43.503Z]       /usr/local/go/src/math/rand/rand.go:196 +0xc2
[2021-10-27T02:27:43.503Z]   github.com/pingcap/tidb/expression.(*decimalJSONGener).gen()
[2021-10-27T02:27:43.503Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/bench_test.go:360 +0x45
[2021-10-27T02:27:43.503Z]   github.com/pingcap/tidb/expression.fillColumnWithGener()
[2021-10-27T02:27:43.503Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/bench_test.go:1005 +0x158
[2021-10-27T02:27:43.503Z]   github.com/pingcap/tidb/expression.fillColumn()
[2021-10-27T02:27:43.503Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/bench_test.go:993 +0x104
[2021-10-27T02:27:43.503Z]   github.com/pingcap/tidb/expression.genVecExprBenchCase()
[2021-10-27T02:27:43.503Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/bench_test.go:1082 +0x438
[2021-10-27T02:27:43.503Z]   github.com/pingcap/tidb/expression.testVectorizedEvalOneVec()
[2021-10-27T02:27:43.504Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/bench_test.go:1107 +0x199
[2021-10-27T02:27:43.504Z]   github.com/pingcap/tidb/expression.TestVectorizedBuiltinCastEvalOneVec()
[2021-10-27T02:27:43.504Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/builtin_cast_vec_test.go:153 +0x54
[2021-10-27T02:27:43.504Z]   testing.tRunner()
[2021-10-27T02:27:43.504Z]       /usr/local/go/src/testing/testing.go:1193 +0x202
[2021-10-27T02:27:43.504Z] 
[2021-10-27T02:27:43.504Z] Previous write at 0x00c000320000 by goroutine 122:
[2021-10-27T02:27:43.504Z]   math/rand.(*rngSource).Uint64()
[2021-10-27T02:27:43.504Z]       /usr/local/go/src/math/rand/rng.go:239 +0x54
[2021-10-27T02:27:43.504Z]   math/rand.(*rngSource).Int63()
[2021-10-27T02:27:43.504Z]       /usr/local/go/src/math/rand/rng.go:234 +0x1d9
[2021-10-27T02:27:43.504Z]   math/rand.(*Rand).Int63()
[2021-10-27T02:27:43.504Z]       /usr/local/go/src/math/rand/rand.go:85 +0xbc
[2021-10-27T02:27:43.504Z]   math/rand.(*Rand).Float64()
[2021-10-27T02:27:43.504Z]       /usr/local/go/src/math/rand/rand.go:196 +0xc2
[2021-10-27T02:27:43.504Z]   github.com/pingcap/tidb/expression.(*decimalJSONGener).gen()
[2021-10-27T02:27:43.504Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/bench_test.go:360 +0x45
[2021-10-27T02:27:43.504Z]   github.com/pingcap/tidb/expression.fillColumnWithGener()
[2021-10-27T02:27:43.504Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/bench_test.go:1005 +0x158
[2021-10-27T02:27:43.504Z]   github.com/pingcap/tidb/expression.fillColumn()
[2021-10-27T02:27:43.504Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/bench_test.go:993 +0x104
[2021-10-27T02:27:43.504Z]   github.com/pingcap/tidb/expression.genVecBuiltinFuncBenchCase()
[2021-10-27T02:27:43.504Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/bench_test.go:1223 +0x417
[2021-10-27T02:27:43.504Z]   github.com/pingcap/tidb/expression.testVectorizedBuiltinFunc()
[2021-10-27T02:27:43.504Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/bench_test.go:1344 +0x5a4
[2021-10-27T02:27:43.504Z]   github.com/pingcap/tidb/expression.TestVectorizedBuiltinCastFunc()
[2021-10-27T02:27:43.504Z]       /home/jenkins/agent/workspace/tidb-unit-test-nightly/go/src/github.com/pingcap/tidb/expression/builtin_cast_vec_test.go:157 +0x54
[2021-10-27T02:27:43.504Z]   testing.tRunner()
[2021-10-27T02:27:43.504Z]       /usr/local/go/src/testing/testing.go:1193 +0x202
```

### 1. Minimal reproduce step (Required)

in ci https://ci.pingcap.net/blue/rest/organizations/jenkins/pipelines/tidb-unit-test-nightly/runs/13151/nodes/114/steps/461/log/?start=0

### 4. What is your TiDB version? (Required)
master

### LINKED ISSUE #29283: Unstable test: builtin_string_test.go:2634 TestWeightString
## Bug Report

```
[2021-10-31T11:13:28.659Z] --- FAIL: TestWeightString (0.00s)

[2021-10-31T11:13:28.659Z]     builtin_string_test.go:2634: 

[2021-10-31T11:13:28.659Z]         	Error Trace:	builtin_string_test.go:2634

[2021-10-31T11:13:28.659Z]         	Error:      	Not equal: 

[2021-10-31T11:13:28.659Z]         	            	expected: "a "

[2021-10-31T11:13:28.659Z]         	            	actual  : "a"

[2021-10-31T11:13:28.659Z]         	            	

[2021-10-31T11:13:28.659Z]         	            	Diff:

[2021-10-31T11:13:28.659Z]         	            	--- Expected

[2021-10-31T11:13:28.659Z]         	            	+++ Actual

[2021-10-31T11:13:28.659Z]         	            	@@ -1 +1 @@

[2021-10-31T11:13:28.659Z]         	            	-a 

[2021-10-31T11:13:28.659Z]         	            	+a

[2021-10-31T11:13:28.659Z]         	Test:       	TestWeightString
```

See https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_check_2/detail/tidb_ghpr_check_2/42054/pipeline/

### PR DESCRIPTION
**Pull request by tisonkun**
... also use a thread safe rng for stable test

This closes #29132 .
This closes #29134 .
This closes #29142 .

This closes #29283 . Move `TestCIWeightString` to serial test.

This closes #29164 . Replace `rand.Source` with `lockedSource`.

Signed-off-by: tison <wander4096@gmail.com>

### Release note

```release-note
None
```

### CODE REVIEW

File: expression/bench_test.go [hunk 1]
```diff
@@ -218,8 +219,26 @@ type defaultRandGen struct {
 	*rand.Rand
 }
 
+type lockedSource struct {
```
Comment (feitian124) [inline expression/bench_test.go:222]: why add this struct? seems not used in other changed files.

Comment (feitian124) [reply expression/bench_test.go:222]: get it, as some tests are run in parallal, so need make newDefaultRandGen safe for concurrent

**Comment by feitian124** [comment]
LGTM.
just wonder, how do you find some tests must run in serial? some of them are not clear at all.

File: expression/builtin_string_serial_test.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	"unicode/utf8"
+
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/collate"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCIWeightString(t *testing.T) {
```
Comment (unconsolable) [inline expression/builtin_string_serial_test.go:28]: Shall the `TestWeightString` should be moved to here too? it seems that https://github.com/pingcap/tidb/issues/29283 happens on `TestWeightString`

Comment (tisonkun) [reply expression/builtin_string_serial_test.go:28]: Nope. The root cause of `TestWeightString` is running concurrently with `TestCIWeightString`. Thus after we move `TestCIWeightString` out, `TestWeightString` will be fixed.

**Comment by tisonkun** [comment]
> /run-common-test tidb-test=pr/1307

what does it run for?

**Comment by winoros** [comment]
> > /run-common-test tidb-test=pr/1307
> 
> what does it run for?

Select a pr to test the changes(

**Comment by tisonkun** [comment]
> > > /run-common-test tidb-test=pr/1307
> > 
> > 
> > what does it run for?
> 
> Select a pr to test the changes(

It may confuse other reviewers that this PR has a bug =。=

**Review by bb7133 — APPROVED** [review]

**Pull request merged**